hail/hail/ir-gen/src/Main.scala

```diff
@@ -0,0 +1,1239 @@
+      attsAndChildren.foldLeft[SeqRepr[Child]](SeqRepr.empty(Child_("BaseIR")))(
+        _ ++ _.childrenSeq
+      ).toString
+
+    private def childrenOffsets: Seq[IntRepr] =
+      attsAndChildren.scanLeft(IntRepr())(_ + _.nChildren)
+
+    private def copyMethod: String = {
+      val decl = s"override def copyWithNewChildren(newChildren: IndexedSeq[BaseIR]): $name = "
+      val assertion = s"assert(newChildren.length == $nChildren)"
+      val body = name + (attsAndChildren, childrenOffsets).zipped.map { case (x, offset) =>
+        val newChildren = SeqRepr.Dynamic("newChildren", Child_("BaseIR"))
+        x.repr.copyWithNewChildren(
+          if (x.nChildren.hasStaticValue(0)) SeqRepr.empty(Child_("BaseIR"))
+          else if (x.nChildren.hasStaticValue(1))
+            SeqRepr.Static(Seq(newChildren(offset)), Child_("BaseIR"))
+          else if (x.nChildren == nChildren) newChildren
+          else newChildren.slice(offset, x.nChildren)
+        )
+      }.mkString("(", ", ", ")")
+      decl + "{\n    " + assertion + "\n    " + body + "}"
+    }
+
+    private def classDecl = {
+      val paramList = s"$name(${attsAndChildren.map(_.generateDeclaration).mkString(", ")})"
+      s"final case class $paramList extends IR" + traits.map(" with " + _.name).mkString
+    }
+
+    private def classBody = {
+      val childrenSeqDef = if (nChildren.hasStaticValue(0))
+        s"override def childrenSeq: IndexedSeq[BaseIR] = IndexedSeq.empty"
+      else
+        s"override lazy val childrenSeq: IndexedSeq[BaseIR] = $children"
+      val extraMethods =
+        this.extraMethods :+ childrenSeqDef :+ copyMethod
+      val constraints = this.constraints ++ attsAndChildren.flatMap(_.constraints)
+      if (constraints.nonEmpty || extraMethods.nonEmpty) {
+        (
+          " {" +
+            (if (constraints.nonEmpty)
+               constraints.map(c => s"  require($c)").mkString("\n", "\n", "\n")
+             else "")
+            + (
+              if (extraMethods.nonEmpty)
+                extraMethods.map("  " + _).mkString("\n", "\n", "\n")
+              else ""
+            )
+            + "}"
+        )
+      } else ""
+    }
+
+    private def classDef =
+      (if (docstring.nonEmpty) s"\n/** $docstring*/\n" else "") + classDecl + classBody
+
+    private def companionDef = if (hasCompanionExtension)
+      s"object $name extends ${name}CompanionExt\n"
+    else ""
+
+    def generateDef: String = companionDef + classDef + "\n"
+  }
+}
+
+object Main {
+  val irdsl: IRDSL = IRDSL_Impl
+  import irdsl._
+
+  private val errorID = in("errorID", att("Int")).withDefault("ErrorIDs.NO_ERROR")
+
+  private def _typ(t: String = "Type") = in("_typ", att(t))
+
+  private val mmPerElt =
+    in("requiresMemoryManagementPerElement", att("Boolean")).withDefault("false")
+
+  private def allNodes: Seq[NodeDef] = {
+
+    val r = Seq.newBuilder[NodeDef]
+
+    r += node("I32", in("x", att("Int"))).withTraits(TrivialIR)
```

Comment:
A nice addition would be to make all these `TypedIR` to avoid casting.
Feel free to dismiss or defer to a later change.